File: code/web/src/api/auth.api.ts
import api from '@/api/axios';
import type {
  ApiResponse,
  LoginRequest,
  LoginResponse,
  VerifyOtpResponse,
  SetupOtpResponse,
  ConfirmOtpSetupResponse,
  RefreshResponse,
} from '@/types/auth.types';

export async function login(
  username: string,
  password: string,
): Promise<ApiResponse<LoginResponse>> {
  const payload: LoginRequest = { username, password };
  const response = await api.post<ApiResponse<LoginResponse>>(
    '/auth/login',
    payload,
  );
  return response.data;
}

export async function verifyOtp(
  otpCode: string,
  tempToken: string,
): Promise<ApiResponse<VerifyOtpResponse>> {
  const response = await api.post<ApiResponse<VerifyOtpResponse>>(
    '/auth/verify-otp',
    { otpCode },
    { headers: { Authorization: `Bearer ${tempToken}` } },
  );
  return response.data;
}

export async function setupOtp(): Promise<ApiResponse<SetupOtpResponse>> {
  const response = await api.post<ApiResponse<SetupOtpResponse>>(
    '/auth/setup-otp',
  );
  return response.data;
}

export async function confirmOtpSetup(
  otpCode: string,
): Promise<ApiResponse<ConfirmOtpSetupResponse>> {
  const response = await api.post<ApiResponse<ConfirmOtpSetupResponse>>(
    '/auth/confirm-otp-setup',
    { otpCode },
  );
  return response.data;
}

export async function refreshToken(
  token: string,
): Promise<ApiResponse<RefreshResponse>> {
  const response = await api.post<ApiResponse<RefreshResponse>>(
    '/auth/refresh',
    { refreshToken: token },
  );
  return response.data;
}

export async function logout(
  refreshTokenValue: string,
): Promise<ApiResponse<null>> {
  const response = await api.post<ApiResponse<null>>('/auth/logout', {
    refreshToken: refreshTokenValue,
  });
  return response.data;
}

File: code/mobile/app/src/main/java/com/example/smtts/data/api/AuthApi.kt
package com.example.smtts.data.api

import com.example.smtts.data.model.ActivateRequest
import com.example.smtts.data.model.ActivateResponse
import com.example.smtts.data.model.ApiResponse
import com.example.smtts.data.model.ChangePasswordRequest
import com.example.smtts.data.model.ChangePasswordResponse
import com.example.smtts.data.model.LoginRequest
import com.example.smtts.data.model.LoginResponse
import com.example.smtts.data.model.RefreshRequest
import com.example.smtts.data.model.RefreshResponse
import com.example.smtts.data.model.VerifyOtpRequest
import com.example.smtts.data.model.VerifyOtpResponse
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.Header
import retrofit2.http.POST

interface AuthApi {

    @POST("auth/login")
    suspend fun login(@Body request: LoginRequest): Response<ApiResponse<LoginResponse>>

    @POST("auth/activate")
    suspend fun activate(@Body request: ActivateRequest): Response<ApiResponse<ActivateResponse>>

    @POST("auth/verify-otp")
    suspend fun verifyOtp(
        @Body request: VerifyOtpRequest,
        @Header("Authorization") tempToken: String
    ): Response<ApiResponse<VerifyOtpResponse>>

    @POST("auth/refresh")
    suspend fun refreshToken(@Body request: RefreshRequest): Response<ApiResponse<RefreshResponse>>

    @POST("auth/logout")
    suspend fun logout(@Body request: RefreshRequest): Response<ApiResponse<Unit>>

    @POST("auth/change-password")
    suspend fun changePassword(@Body request: ChangePasswordRequest): Response<ApiResponse<ChangePasswordResponse>>
}
